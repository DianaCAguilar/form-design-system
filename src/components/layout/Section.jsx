import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

const VALID_BG_NORMAL = ['white', 'haze', 'lightGray'];

const VALID_BG_INVERTED = ['navy'];

const VALID_PADDING = ['default', 'double', 'half', 'none'];

/**
 * @param {String} axis 'h' or 'v'
 * @param {String} amount
 * @returns {String} padding utility classes
 */
export const getPaddingClasses = (axis, amount) => {
  const sides = axis === 'h' ? ['left', 'right'] : ['top', 'bottom'];

  let classes = `padding--${sides[0]} padding--${sides[1]}`;
  if (amount !== 'default') {
    classes = `padding--${sides[0]}--${amount} padding--${sides[1]}--${amount}`;
  }

  return classes;
};

/**
 * @param {String} direction
 * @returns {String} border utility classes
 */
export const getBorderClasses = (direction) => {
  let classes = '';
  switch (direction) {
    case 'h':
      classes = 'border--left border--right';
      break;
    case 'v':
      classes = 'border--top border--bottom';
      break;
    default:
      classes = `border--${direction}`;
  }
  return classes;
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Section = ({ hPadding, vPadding, bgColor, border, children }) => {
  const classNames = cx(
    `bgColor--${bgColor}`,
    'display--block',
    getPaddingClasses('h', hPadding),
    getPaddingClasses('v', vPadding),
    {
      inverted: VALID_BG_INVERTED.includes(bgColor),
      [getBorderClasses(border)]: Boolean(border),
    }
  );
  const styles = {
    backgroundClip: 'border-box',
  };

  return (
    <div style={styles} className={classNames}>
      {children}
    </div>
  );
};

Section.defaultProps = {
  hPadding: 'default',
  vPadding: 'default',
  bgColor: 'white',
};

Section.propTypes = {
  /** React children */
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.element]),

  /** Padding for left and right sides */
  hPadding: PropTypes.oneOf(VALID_PADDING),

  /** Padding for top and bottom sides */
  vPadding: PropTypes.oneOf(VALID_PADDING),

  /**
   * Background color.
   * Type classes (e.g. color--primary)  will automatically invert for dark backgrounds.
   */
  bgColor: PropTypes.oneOf([...VALID_BG_NORMAL, ...VALID_BG_INVERTED]),

  /**
   * Adds border to given side(s).
   * `h` adds border to both left and right.
   * `v` adds border to both top and bottom.
   */
  border: PropTypes.oneOf(['top', 'right', 'bottom', 'left', 'h', 'v', 'all']),
};

export default Section;
