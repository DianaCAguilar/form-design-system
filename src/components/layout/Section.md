Section is a layout component that can be used for sectioning content either vertically or
horizontally.

## Responsibilities of <code>Section</code>

* applying standard background colors
* inverting text for dark backgrounds
* applying standard padding values
* applying standard borders

## Limitations

Padding must be specified by axis. `Section` does not accept padding for specific sides. 

In other words, this component enforces that the top/bottom or the left/right padding
always match.

