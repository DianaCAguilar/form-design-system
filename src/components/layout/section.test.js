import React from 'react';
import { shallow } from 'enzyme';

import Section, { getPaddingClasses, getBorderClasses } from './Section';

const renderComponent = (props) => shallow(<Section {...props}>Content</Section>);

describe('Section component', () => {

  it('matches snapshot (default props)', () => {
    const section = renderComponent();
    expect(section).toMatchSnapshot();
  });

  it('matches snapshot (all props set to user value)', () => {
    /* should also include `inverted` class for navy sections */
    const section = renderComponent({
      hPadding: 'double',
      vPadding: 'half',
      bgColor: 'navy',
      border: 'y',
    });
    expect(section).toMatchSnapshot();
  });

  describe('getPaddingClasses', () => {
    it('handles default padding', () => {
      const actual = getPaddingClasses('x', 'default');
      const expected = 'padding--left padding--right';
      expect(actual).toBe(expected);
    });
    it('handles double padding', () => {
      const actual = getPaddingClasses('y', 'double');
      const expected = 'padding--top--double padding--bottom--double';
      expect(actual).toBe(expected);
    });
  })

  describe('getBorderClasses', () => {
    it('sets border class for a specific direction', () => {
      const actual = getBorderClasses('top');
      const expected = 'border--top';
      expect(actual).toBe(expected);
    });
    it('sets border class for "all"', () => {
      const actual = getBorderClasses('all');
      const expected = 'border--all';
      expect(actual).toBe(expected);
    });
    it('sets border class for "h" sides', () => {
      const actual = getBorderClasses('x');
      const expected = 'border--left border--right';
      expect(actual).toBe(expected);
    });
    it('sets border class for "v" sides', () => {
      const actual = getBorderClasses('y');
      const expected = 'border--top border--bottom';
      expect(actual).toBe(expected);
    });
  })

});
