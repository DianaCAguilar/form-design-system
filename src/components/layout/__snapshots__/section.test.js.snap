// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section component matches snapshot (all props set to user value) 1`] = `
<div
  className="bgColor--navy display--block padding--left--double padding--right--double padding--top--half padding--bottom--half inverted border--top border--bottom"
  style={
    Object {
      "backgroundClip": "border-box",
    }
  }
>
  Content
</div>
`;

exports[`Section component matches snapshot (default props) 1`] = `
<div
  className="bgColor--white display--block padding--left padding--right padding--top padding--bottom"
  style={
    Object {
      "backgroundClip": "border-box",
    }
  }
>
  Content
</div>
`;
